sphinx/writers/html.py

```diff
@@ -76,7 +77,23 @@ def translate(self) -> None:
                      'footer', 'html_prolog', 'html_head', 'html_title',
                      'html_subtitle', 'html_body', ):
             setattr(self, attr, getattr(visitor, attr, None))
-        self.clean_meta = ''.join(self.visitor.meta[2:])
+        self.clean_meta = self.get_clean_meta()
+
+    def get_clean_meta(self) -> str:
+        visitor = cast(HTMLTranslator, self.visitor)
+        meta = visitor.meta[:]
+
+        # Remove <meta charset> tag from docutils
+        content_type = visitor.content_type % visitor.settings.output_encoding
```

Comment:
Given the original report (our use case) is a translator which specifically does not inherit from docutils' HTMLTranslator, I'd say there's no guarantee these attributes will be present.

I can see a few possibilities / alternatives here:

1. check if the visitor is an instance of docutils' HTMLTranslator and only process the items in that case, however if third-parties have implemented a translator which doesn't extend docutils' but sets these it's going to fail
2. check if the *attributes* are present and only strip the attributes in that case (this could handle translators which replicate more of the original behaviour than ours do), however in a case similar to (1) third parties might have set the metas to remove without the corresponding docutils' attributes
3. specifically parse and process the `charset` and `generator` metas, either to strip them entirely or to fix them using sphinx semantics instead (replace `charset="unicode"` by the value of `html_output_encoding`, and mark Sphinx as the [generator](https://html.spec.whatwg.org/multipage/semantics.html#meta-generator) -- at lest if it finds docutils there
4. move the meta-cleaning operation to an (optional) feature of the translator e.g. `self.clean_meta = ''.join(getattr(visitor, 'get_clean_meta', visitor.meta))`, this way the translator has an option to provide clean metas (third parties might see unexpected "padding" metas reappear in the output but that seems like a somewhat lower risk)

I'd favor 3 then 4 personally, the first two seem kinda brittle. Maybe @birkenfeld remembers why they straight stripped the metas created by docutils in the first place though? That might be useful insight.

Reply:
I noticed Sphinx's HTML builder is already strongly coupled with docutils' HTMLTranslator. It uses `astext()`, `meta` and etc. attributes.
https://github.com/sphinx-doc/sphinx/blob/3336dcc2aea0cc93d54ba661d4da65ea46f088b0/sphinx/writers/html.py#L72-L79

Reply:
> I noticed Sphinx's HTML builder is already strongly coupled with docutils' HTMLTranslator. It uses `astext()`, `meta` and etc. attributes.
> 
> https://github.com/sphinx-doc/sphinx/blob/3336dcc2aea0cc93d54ba661d4da65ea46f088b0/sphinx/writers/html.py#L72-L79

To a limited extent, it does have a few attributes and methods it expects but these are relatively easy to implement without extending docutils' HTMLTranslator. I'm not saying they're super decoupled, but it's really quite easy to get a working translator from just the nodevisitor, the serialisation of the meta list is really the only bit which had me stumped.